### ISSUE TITLE
Puzzling behavior of `.[]|=empty`

### ISSUE THREAD

**Issue by zstadler** [stedolan/jq#1771]
As far as I understand, the use of `empty` in update-assignment `|=` deletes the entry, similar to `del(path)`, and `del(path)` also works in arrays.

I tried to selectively delete and modify array entries, and got strange results. While trying to understand the problem, I minimized the code to this example, which I fail to understand its behavior:
```
jq -n -c "{data: [0,1,2,3,4,5]} | .data[] |= empty"
{"data":[1,3,5]}
```
Version 1.5 gives a different puzzling result:
```
jq-1.5 -n -c "{data: [0,1,2,3,4,5]} | .data[] |= empty"
null
```

See also an example similar to the one given [on stackoverflow](https://stackoverflow.com/questions/53381867/puzzling-behavior-of-empty):
```
jq -n -c "{data: [0,1,2,3,4,5]}| .data[] |= if . == 2 then empty else . end"
{"data":[0,1,3,4,5,null]}
```
and its 1.5 version result:
```
jq-1.5 -n -c "{data: [0,1,2,3,4,5]}| .data[] |= if . == 2 then empty else . end"
{"data":[null,null,null,null,null,null]}
```

**Comment by LordMike** [comment]
https://stedolan.github.io/jq/manual/#Update-assignment:|=

There was a change in 1.5, as mentioned, but it's funny it also affects the `empty` case. Perhaps taking the first element of a list is related to that?

**Comment by nicowilliams** [comment]
Long story short:

- assignments are reductions
- reductions cannot keep a copy of the reduction state across the evaluation of the reduction state update operation, as this can cause excessive allocation (see below about why)
- therefore there's nothing good that can be done when a reduction state update expression produces no values (or even many values, for that matter)

The reason that holding a reference to the reduction state across the reduction state update expression evaluation causes excessive allocation is simple: jq's internal (`jv`) data structures are immutable, copy-on-write, but when a `jv` has just `1` reference to it, then mutations do not copy but actually mutate in-situ.  Therefore, not keeping a reference to the reduction state across the evaluation of the state update expression means that after the first update the reduction state always has just one reference, which means that all subsequent evaluations of the state update expression can mutate the state without unnecessary allocations and copies.  Conversely, holding that reference during the update would mean that every evaluation of the state update expression must allocate and copy.  This is bad, and that's why the reduction operator does not hold a reference to the reduction state during evaluation of the state update expression.  And because assignments are reductions...

The only reasonable behaviors (from an implementation point of view) in the face of `empty` state updates are: a) raise an `error`, or b) `break` out of the reduction operation, or maybe c) use some value as the next reduction state?? `null`, or perhaps re-initialize the reduction..., or maybe d) terminate the reduction but output `null` or some such.  All of these are going to be surprising to users, except perhaps for (a).

Oh, I suppose there is one more thing we could do: we could have `empty` (really, the `BACKTRACK` opcode) behave specially in a reduction, since its input, `.` could be preserved as the next reduction state.  But this too could have surprises, as `empty`'s `.` might not be the same as the state update expression's `.`.

So, in short, don't do this.

The workaround is simple: use `first(exp, .)` if you don't mind the extra allocations, or `first(exp, "dur"|error)` if you want an error, or `first(exp, some_default_state)`, or similar.

**Comment by zstadler** [comment]
I understand that this type of assignment will not be handled by the program.

If such user-errors cannot be detected at compile time, then jq should raise an `error` at run-time.

A user-oriented explanation of the limitation would also be greatly appreciated as part of  https://stedolan.github.io/jq/manual/#Update-assignment:|=
For example, is the problem related only to assignment of `empty` values? Is it related only to the `[]` operator? etc.

**Comment by zstadler** [comment]
BTW, my preferred workaround is to use `map`. For example,
```
jq -n -c "{data: [0,1,2,3,4,5]}| .data |= map( if . == 2 then empty else . end)"
{"data":[0,1,3,4,5]}
```
instead of
```
jq -n -c "{data: [0,1,2,3,4,5]}| .data[] |= if . == 2 then empty else . end"
```
